pennylane/workflow/set_shots.py

```diff
@@ -16,75 +16,148 @@
 """
 from __future__ import annotations
 
-from collections.abc import Callable, Sequence
-from typing import TYPE_CHECKING, Optional
+from collections.abc import Callable
+from typing import TYPE_CHECKING, overload
 
 from .qnode import QNode
 
-# Sentinel value to distinguish decorator mode from direct calls
-_DECORATOR_MODE = object()
-
 if TYPE_CHECKING:
-    from pennylane.measurements import Shots
+    from pennylane.measurements import ShotsLike
+
+
+# Sentinel value to detect when shots parameter is not provided
+_SHOTS_NOT_PROVIDED = object()
+
 
+@overload
+def set_shots(qnode: QNode, shots: ShotsLike) -> QNode: ...
 
+
+@overload
+def set_shots(shots: ShotsLike) -> Callable[[QNode], QNode]: ...
+
+
+@overload
 def set_shots(
-    qnode: QNode | object = _DECORATOR_MODE,
-    shots: Optional[Shots | int | Sequence[int | tuple[int, int]]] = None,
-) -> QNode | Callable[[QNode], QNode]:
-    """Transform used to set or update a circuit's shots.
+    *, shots: ShotsLike
+) -> Callable[[QNode], QNode]: ...
+
+
+def set_shots(*args, shots: ShotsLike = _SHOTS_NOT_PROVIDED):
+    """Transform used to set or update a circuit's shots."""
+    # Keyword-only case: @set_shots(shots=500) or @set_shots(shots=None)
+    if len(args) == 0 and shots is not _SHOTS_NOT_PROVIDED:
+        return _create_keyword_decorator(shots)
+
+    if len(args) == 1 and shots is not _SHOTS_NOT_PROVIDED:
+        # Direct application: set_shots(qnode, shots=500) or set_shots(qnode, shots=None)
+        return _apply_shots_to_qnode(args[0], shots)
+    if len(args) == 1 and shots is _SHOTS_NOT_PROVIDED:
+        # Positional decorator: @set_shots(500) or @set_shots(None)
+        return _set_shots_dispatch(args[0])
+
+    raise ValueError(f"Invalid arguments to set_shots: {args=}, {shots=}")
+
+
+def _set_shots_dispatch(shots_value: ShotsLike) -> Callable[[QNode], QNode]:
+    """Default case: @set_shots(500) - positional shots value"""
+
+    def positional_decorator(qnode_func: QNode) -> QNode:
+        return qnode_func.update_shots(shots_value)
```

Comment:
Should we validate that `qnode_func` is indeed a `QNode`?

Reply:
https://github.com/PennyLaneAI/pennylane/pull/7919/commits/715caa89b6f86f726bcf23a989974cc2fe2743e4